pyro/contrib/gp/models/gplvm.py

```diff
@@ -0,0 +1,114 @@
+from pyro.params import param_with_module_name
+
+
+class GPLVM(Parameterized):
+    """
+    Gaussian Process Latent Variable Model (GPLVM) model.
+
+    GPLVM is a Gaussian Process model with its train input data is a latent variable.
+    This model is useful for dimensional reduction of high dimensional data. Assume the
+    mapping from low dimensional latent variable to is a Gaussian Process instance.
+    Then the high dimensional data will play the role of train output ``y`` and our
+    target is to learn latent inputs which best explain ``y``. For the purpose of
+    dimensional reduction, latent inputs should have lower dimensions than ``y``.
+
+    We follows reference [1] to put a unit Gaussian prior to the input and approximate
+    its posterior by a multivariate normal distribution with two variational
+    parameters: ``X_loc`` and ``X_scale_tril``.
+
+    Reference:
+
+    [1] Bayesian Gaussian Process Latent Variable Model
+    Michalis K. Titsias, Neil D. Lawrence
+
+    :param ~pyro.contrib.gp.models.model.GPModel base_model: A Pyro Gaussian Process
+        model object. Note that ``base_model.X`` will be the initial value for the
+        variational parameter ``X_loc``.
+    :param str name: Name of this model.
+    """
+    def __init__(self, base_model, name="GPLVM"):
+        super(GPLVM, self).__init__(name)
+        if base_model.X.dim() != 2:
+            raise ValueError("GPLVM model only works with 2D latent X, but got "
+                             "X.dim() = {}.".format(base_model.X.dim()))
+        self.base_model = base_model
+        self.y = self.base_model.y
+
+        self.X_loc = Parameter(self.base_model.X)
+
+        C = self.X_loc.shape[1]
+        X_scale_tril_shape = self.X_loc.shape + (C,)
+        Id = torch.eye(C, out=self.X_loc.new_empty(C, C))
+        X_scale_tril = Id.expand(X_scale_tril_shape)
+        self.X_scale_tril = Parameter(X_scale_tril)
+        self.set_constraint("X_scale_tril", constraints.lower_cholesky)
+
+        self._call_base_model_guide = True
+
+    def model(self):
+        self.set_mode("model", only_this_module=True)
+
+        zero_loc = self.X_loc.new_zeros(self.X_loc.shape)
+        C = self.X_loc.shape[1]
+        Id = torch.eye(C, out=self.X_loc.new_empty(C, C))
+        X_name = param_with_module_name(self.name, "X")
+        X = pyro.sample(X_name, dist.MultivariateNormal(zero_loc, scale_tril=Id)
+                                    .independent(zero_loc.dim()-1))
+
+        self.base_model.set_data(X, self.y)
+        self.base_model.model()
+
+    def guide(self):
+        self.set_mode("guide", only_this_module=True)
+
+        X_loc = self.get_param("X_loc")
+        X_scale_tril = self.get_param("X_scale_tril")
+        X_name = param_with_module_name(self.name, "X")
+        X = pyro.sample(X_name,
+                        dist.MultivariateNormal(X_loc, scale_tril=X_scale_tril)
+                            .independent(X_loc.dim()-1))
```

Comment:
Here we define `guide` for X. The only difference to other GP models is we allow X to be a parameter (with unit normal prior and variational guide (not the MAP guide because according to literature, MAP guide produces poor results)).